sw/device/lib/dif/README.md

```diff
@@ -74,26 +93,121 @@ there are some relaxations of these rules for them described at the end.
+  point to initialized data.
+* `result_t dif_<ip>_configure(const dif_<ip>_t *handle, dif_<ip>_config_t
+  config);` is a function that configures the hardware managed by `handle` with
+  runtime parameters in an implementation-defined way. This function should be
+  "one-off": it should only need to be called once for the lifetime of the handle.
+
+  If there is no meaningful state to configure, this function may be omitted.
+
+#### Interrupt Management
+
+The following types and functions are a standard inteface for peripherals that
+provide a collection of `INTR_ENABLE`, `INTR_STATE`, and `INTR_TEST` registers
+for interrupt management. A DIF library for a peripheral providing such
+registers must provide this interface.
+
+If a peripheral is defined with `no_auto_intr_regs: true`, this exact API is not
+required even of the above registers are provided (though DIF libraries are
+encouraged to follow it where it makes sense).
+
+* `dif_<ip>_irq_t` is an enum that lists all of the interrupt types for this
+  peripheral. A DIF library may opt to use another pre-existing type instead,
+  when, for example, interrupt types are coupled to a distinct
+  peripheral-specific concept. In that case, all occurences of `dif_<ip>_irq_t`
+  belkow should be replaced with this typoe.
+* `result_t dif_<ip>_irq_is_pending(const dif_<ip>_t *handle, dif_<ip>_irq_t
+  irq_type, bool *is_pending);` checks whether a specific interrupt is
+  pending (i.e., if the interrupt has been asserted but not yet cleared).
+* `result_t dif_<ip>_irq_clear(const dif_<ip>_t *handle, dif_<ip>_irq_t
+  irq_type);` clears a specific interrupt (and does nothing if the interrupt is
+  not currrently pending).
+* `result_t dif_<ip>_irq_get_enabled(const dif_<ip>_t *handle, dif_<ip>_irq_t
+  irq_type, dif_<ip>_toggle_t *state);` gets whether an interrupt is enabled
+  (i.e., masked).
+* `result_t dif_<ip>_irq_set_enabled(const dif_<ip>_t *handle, dif_<ip>_irq_t
+  irq_type, dif_<ip>_toggle_t state);` sets whether a particular interrupt is
+  enabled (i.e., masked).
+* `result_t dif_<ip>_irq_force(const dif_<ip>_t *handle, dif_<ip>_irq_t
+  irq_type);` forcibly asserts a specific interrupt, causing it to be serviced
+  as if hardware had triggered it.
+
+Additionally, the following types allow for batch save/restore operations on
+the interrupt enable register:
+
+* `dif_<ip>_irq_snapshot_t` is a type that encapsulates restorable interrupt
+  state, to be used with the two functions below. This type should be treated as
+  opaque by clients.
+* `result_t dif_<ip>_irq_disable_all(const dif_<ip>_t *handle,
+  dif_<ip>_irq_snapshot_t *snapshot);` disables all interrupts associated with
+  the peripheral, saving them to `*snapshot`. `snapshot` may be null, in which
+  case the previous enablement state is not saved.
+* `result_t dif_<ip>_irq_restore_all(const dif_<ip>_t *handle,
+  dif_<ip>_irq_snapshot_t *snapshot);` restores an interrupt enablement snapshot
```

Comment:
`const dif_<ip>_irq_snapshot_t *snapshot` or pass by value? IIUC `*snapshot` is not an out-parameter in this case. Looking at the template code bellow this should be pass by value.

Reply:
Some comment elsewhere suggested we take the snapshot by pointer, for symmetry with disable_all(). I thought this was a reasonable suggestion, seeing as the type is opaque.

Reply:
We should have a `const` here.